File: Leetcode/merge_k_sorted_lists/merge_k_sorted_lists_v2.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

import heapq

class Solution:
    def mergeKLists(self, lists: list[ListNode]) -> ListNode:

        # Initialize sorted list head
        head = ListNode()
        curr = head

        # K elements of the curr node of each list will be stored in a min hip, at each traversal an element from list[i] will transfer from minheap to sorted list (Time Complexity = O(log(k))). Next element of list[i] will be added to the sorted list (Time Complexity = O(log(k))).
        heap = []
        heapq.heapify(heap)

        k = len(lists)

        for i in range(k):
            if lists[i]:
                heapq.heappush(heap, (lists[i].val, i))
                lists[i] = lists[i].next
        
        # Till the heap is empty, i.e. all elements of all lists are inserted
        while len(heap)>0:
            val, listid = heapq.heappop(heap)
            # Add min elem of heap to sorted list
            curr.next = ListNode(val)
            curr = curr.next
            if lists[listid]:
                heapq.heappush(heap, (lists[listid].val, listid))
                lists[listid] = lists[listid].next
        
        return head.next

File: Leetcode/merge_k_sorted_lists/merge_k_sorted_lists_v2.cpp
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;
/**
 * Definition for singly-linked list.
 */ 
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:

    struct CompareNodes {
        bool operator()(const ListNode* a, const ListNode* b) {
            return a->val > b->val;
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, CompareNodes> pq;
        for (ListNode* list : lists) {
            if (list) {
                pq.push(list);
            }
        } 

        ListNode* dummy = new ListNode();
        ListNode* current = dummy;

        while (!pq.empty()) {
            ListNode* smallest = pq.top();
            pq.pop();

            current->next = smallest;
            current = current->next;

            if (smallest->next) {
                pq.push(smallest->next);
            }
        }
        return dummy->next;
    }
};